src/Language/PureScript/TypeChecker/Types.hs

```diff
@@ -463,6 +544,59 @@ infer' (TypedValue checkType val ty) = do
   ty' <- introduceSkolemScope <=< replaceAllTypeSynonyms <=< replaceTypeWildcards $ elabTy
   tv <- if checkType then withScopedTypeVars moduleName args (check val ty') else return (TypedValue' False val ty)
   return $ TypedValue' True (tvToExpr tv) ty'
+infer' (VisibleTypeApp val arg@(TypeWildcard _ _)) = do
+  TypedValue' _ val' valTy <- infer' val
+  case removeFirstVisible valTy of
+    Just valTy' ->
+      pure $ TypedValue' True val' valTy'
+    Nothing ->
+      throwError . errorMessage $ CannotApplyExpressionOfTypeOnType valTy arg
+infer' (VisibleTypeApp val arg) = do
+  TypedValue' _ val' valTy <- infer' val
+  arg' <- introduceSkolemScope <=< replaceAllTypeSynonyms <=< replaceTypeWildcards $ arg
+  (arg'', argKnd) <- kindOf arg'
+  case partitionQuantifiers valTy of
+    -- bfrVis: quantified variables before the visible variable
+    -- visArg: the visible quantified variable
+    -- visKnd: the visible quantified variable's kind
+    -- aftVis: quantified variables after the visible variable
+    -- valTy': the underlying monotype stripped of quantifiers
+    Just (bfrVis, (_, visArg, visKnd, _, _), aftVis, valTy') -> do
+      let
+        -- Collect all free type variables in visKnd, such that
+        -- quantifiers in bfrVis can be partitioned by whether
+        -- or not they appear within visKnd.
+        --
+        -- Given:
+        --
+        -- forall (j :: Type) (k :: Type) (@t :: k -> Type)
+        --
+        -- Then:
+        --
+        -- 1. freeInVisKnd: { k }
+        -- 2. bfrVisY: [ k ]
+        -- 3. bfrVisN: [ j ]
```

Comment:
This doesn't take into account fringe cases such as:
```hs
forall (a :: Type) (b :: a) (@c :: Proxy b -> Type)
```
Where `a` also has to be instantiated alongside `b`. I doubt the compiler is ever able to synthesize such a type signature in most regular usage.

If there ever is a need for this, one way to go about it is to build a directed graph to figure out which variables need to be instantiated, starting from the kind of the visible quantifier. For example:
```hs
forall (a :: Type) (e :: Type) (b :: a) (c :: a -> Proxy b) (@d :: Proxy c)
```
Yields something along the lines of:
```hs
[ (d, [c]), (c, [a, b]), (b, [a]), (a, []) (e, []) ]
```
All that's left is to create the unification variables and replace as necessary.

Reply:
Maybe this should be inlined into the code rather than this issue, then?